City-of-Helsinki/hel-fi-drupal-grants
# fix: AU-1922: Refactor handleBankAccountConfirmation and fix for duplicate file names

# [AU-1922](https://helsinkisolutionoffice.atlassian.net/browse/AU-1922)

## What was done
This pull request fixes an issue where an applications bank account confirmation file would disappear if a file with the same name was uploaded to the "Muu liite" section. This pull request also refactors the `handleBankAccountConfirmation` method inside the `AttachmentHandler` class. 

## How to install
Make sure your instance is up and running on the correct branch.
  * `git checkout feature/AU-1922-duplicate-attachment-name-fix`
  * `make fresh`
  * `make drush-cr`

## How to test
Your task is essentially going to be to try to break the attachment implementation of any application. I've done my testing with the [Kaupunginhallitus, yleisavustushakemus](https://hel-fi-drupal-grant-applications.docker.so/fi/uusi-hakemus/yleisavustushakemus) form, since it has a lot of attachment fields. I've tested at least the following things, end encourage you to do as well.

1. Start an application, add a bank account, submit without adding other attachments. Copy the submitted application and check that the new application has the correct confirmation file.

2. Start an application, add a bank account, save as draft, edit and submit without adding other attachments. Copy the submitted application and check that the new application has the correct confirmation file.

3. Start an application, add a bank account, add a file with the same name as the bank account confirmation file to the "Muu liite" section, submit. Copy the submitted application and check that the new application has the correct confirmation file.

4. Start an application, add a bank account, save as draft, edit and add a file with the same name as the bank account confirmation file to the "Muu liite" section, submit. Copy the submitted application and check that the new application has the correct confirmation file.

5. Start an application, add a bank account, save as draft, edit and change bank accounts, submit. Copy the submitted application and check that the new application has the correct confirmation file.

6. Start an application, add a bank account, save as draft, edit and change bank accounts, add a file with the same name as the bank account confirmation file to the "Muu liite" section, submit. Copy the submitted application and check that the new application has the correct confirmation file.

And finally, review the code and make sure we aren't doing something we shouldn't be doing.

[AU-1922]: https://helsinkisolutionoffice.atlassian.net/browse/AU-1922?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
